### PR TITLE
fix: Unnecessary migrations for unsigned numeric types

### LINKED ISSUE #2943: Inappropriate migration generated
**Issue type:**

[ ] question
[ x ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ x ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ x ] `latest` 0.2.7
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

When using an `unsigned` **PrimaryGenerateColumn**, and generating a migration using `typeorm migration:generate`, the generated migrations get into a loop. The generated migration is always the same, because the query leads to an undocumented change in the database, which causes migration:generate to detect the change again and again.

Reproduction:
```ts
import { Entity, PrimaryGeneratedColumn, Column, PrimaryColumn, Generated } from 'typeorm';

/**
 * Log action
 */
@Entity('log_action')
export class LogAction {
   /**
    * logActionId
    */
   @PrimaryGeneratedColumn({
      name: 'log_action_id',
      unsigned: true
   })
   public logActionId: number;

   /**
    * title
    */
   @Column('varchar', {
      nullable: false,
      length: 150,
      name: 'title'
   })
   public title: string;
   /**
    * created
    */
   @Column('datetime', {
      nullable: false,
      default: () => 'CURRENT_TIMESTAMP',
      name: 'created'
   })
   public created: Date;

   /**
    * last_modified
    */
   @Column('datetime', {
      nullable: false,
      default: () => 'CURRENT_TIMESTAMP',
      onUpdate: 'CURRENT_TIMESTAMP',
      name: 'last_modified'
   })
   public lastModified: Date;
}
```

- **Create table**
`ts-node ./node_modules/.bin/typeorm schema:sync`
- **Create migration**
`ts-node ./node_modules/.bin/typeorm migration:generate -n test`

You can run `migration:generate` and `migration:run` multiple times, there will always be this migration generated, again and again.
```ts
import {MigrationInterface, QueryRunner} from "typeorm";

export class test1539781487192 implements MigrationInterface {

    public async up(queryRunner: QueryRunner): Promise<any> {
        await queryRunner.query("ALTER TABLE `log_action` CHANGE `log_action_id` `log_action_id` int UNSIGNED NOT NULL AUTO_INCREMENT");
    }

    public async down(queryRunner: QueryRunner): Promise<any> {
        await queryRunner.query("ALTER TABLE `log_action` CHANGE `log_action_id` `log_action_id` int(10) UNSIGNED NOT NULL AUTO_INCREMENT");
    }

}
```

At this point, it seems to be an issue of not defining a precision, but even adding `precision: 10` to **PrimaryGeneratedColumn** config doesn't solve it. It produces the following migration:
```ts
import { MigrationInterface, QueryRunner } from 'typeorm';

export class test1539782429566 implements MigrationInterface {

   public async up(queryRunner: QueryRunner): Promise<any> {
      await queryRunner.query('ALTER TABLE `log_action` CHANGE `log_action_id` `log_action_id` int(10) UNSIGNED NOT NULL AUTO_INCREMENT');
   }

   public async down(queryRunner: QueryRunner): Promise<any> {
      await queryRunner.query('ALTER TABLE `log_action` CHANGE `log_action_id` `log_action_id` int(10) UNSIGNED NOT NULL AUTO_INCREMENT');
   }

}


```

### PR DESCRIPTION
**Pull request by michaelbromley**
In MariaDB, unsigned numeric types (`<tiny,small,medium>int`) without an explicit `width` property set would generate migrations on every run. This is due to an error in setting the default width for unsigned types.

Fixes #2943

### CODE REVIEW

File: src/query-runner/BaseQueryRunner.ts [hunk 1]
```diff
@@ -324,10 +324,18 @@ export abstract class BaseQueryRunner {
...
 
-        if (this.connection.driver.dataTypeDefaults
+        const defaultWidthForType = this.connection.driver.dataTypeDefaults
             && this.connection.driver.dataTypeDefaults[column.type]
-            && this.connection.driver.dataTypeDefaults[column.type].width) {
-            return this.connection.driver.dataTypeDefaults[column.type].width === width;
+            && this.connection.driver.dataTypeDefaults[column.type].width;
+
+        if (defaultWidthForType) {
+            // the default widths of these numeric types are 1 less when the column is unsigned.
```
Comment (pleerock) [inline src/query-runner/BaseQueryRunner.ts:332]: is this statement true for all databases?

Comment (michaelbromley) [reply src/query-runner/BaseQueryRunner.ts:332]: No, actually only MariaDB. The method this code lives in (`isDefaultColumnWidth`) is only called for the MySQL driver, as stated in the comment above it, and it turns out that the method is actually only hit when using MariaDB specifically, because [this line](https://github.com/typeorm/typeorm/pull/6632/files/414a3faada18a7574e7dfba98006304cb2df0d36#diff-f449ebcfffaeef0bb22211a613c8e283R1301) always results in `width = undefined` for MySQL due to it's representation of `dbColumn["COLUMN_TYPE"]` not including the width part.

Comment (pleerock) [reply src/query-runner/BaseQueryRunner.ts:332]: Having this check for all drivers is confusing. We need to move this logic to a driver-specific method, e.g. `driver.compareColumnWidths` or something like that.

Comment (michaelbromley) [reply src/query-runner/BaseQueryRunner.ts:332]: Makes sense, I agree. I can do this refractor as an addition to this PR.

Ps if you are planning to make a release soon let me know so I can do the changes in time.

Comment (pleerock) [reply src/query-runner/BaseQueryRunner.ts:332]: yeah its already quite a while I didn't make a release, so yeah, expect a release soon

Comment (michaelbromley) [reply src/query-runner/BaseQueryRunner.ts:332]: OK I just looked into this more. The `isDefaultColumnWidth` method is actually used in 2 QueryRunners: MysqlQueryRunner & AuroraQueryRunner. I guess that's why it was put on the base class.

I don't see any tests for Aurora in CI, so I've no way to be sure that my changes are valid for Aurora. 

I question the value of putting it in the Driver interface because as far as I can tell it is an (almost useless) feature of MySQL _only_. [Here's a good explanation](https://alexander.kirk.at/2007/08/24/what-does-size-in-intsize-of-mysql-mean/) of what it actually does.

Here's another suggestion:

1. Move the `isDefaultColumnWidth` method from the Base class to both MysqlQueryRunner and AuroraQueryRunner.
2. Leave the AuroraQueryRunner version as it was before the fix, since I do not know whether it breaks anything there.
3. In the MysqlQueryRunner include the changes from this PR, with an additional check of the existing `isMariaDb` boolean to further make sure (and make explicit) that the fix only applies to MariaDB.

What do you think?

Comment (pleerock) [reply src/query-runner/BaseQueryRunner.ts:332]: yeah, go ahead

**Comment by michaelbromley** [comment]
Strangely that last commit now causes a test failure. Is there any way to tell from CircleCI which DB driver is failing?

**Comment by 0xCAP** [comment]
This would be an awesome fix. I'm getting redundant migrations being generated all the time!

**Comment by pleerock** [comment]
looks like you need to figure it out from the failed SQL

**Comment by michaelbromley** [comment]
OK I figured out the failure is happening with MySQL 5.7, I tested locally with MySQL 8. I'll get this fixed.

**Comment by michaelbromley** [comment]
@pleerock tests all passing, looks good to merge 👍

File: src/driver/mysql/MysqlQueryRunner.ts [hunk 2]
```diff
@@ -1296,6 +1296,7 @@ export class MysqlQueryRunner extends BaseQueryRunner implements QueryRunner {
                     tableColumn.name = dbColumn["COLUMN_NAME"];
                     tableColumn.type = dbColumn["DATA_TYPE"].toLowerCase();
 
+                    tableColumn.unsigned = tableColumn.zerofill ? true : dbColumn["COLUMN_TYPE"].indexOf("unsigned") !== -1;
```
Comment (pleerock) [inline src/driver/mysql/MysqlQueryRunner.ts:1299]: `tableColumn.zerofill` is set below, why did you move it?

File: src/driver/mysql/MysqlQueryRunner.ts [hunk 3]
```diff
@@ -1694,4 +1694,34 @@ export class MysqlQueryRunner extends BaseQueryRunner implements QueryRunner {
...
+
+        const defaultWidthForType = this.connection.driver.dataTypeDefaults
+            && this.connection.driver.dataTypeDefaults[column.type]
+            && this.connection.driver.dataTypeDefaults[column.type].width;
+
+        if (defaultWidthForType) {
+            // In MariaDB & MySQL 5.7, the default widths of certain numeric types are 1 less than
+            // the usual defaults when the column is unsigned.
+            const typesWithReducedUnsignedDefault = ["int", "tinyint", "smallint", "mediumint"];
+            if (column.unsigned && -1 < typesWithReducedUnsignedDefault.indexOf(column.type)) {
```
Comment (pleerock) [inline src/driver/mysql/MysqlQueryRunner.ts:1717]: `column.unsigned && -1` can we replace this check to something more obvious?

File: src/driver/mysql/MysqlQueryRunner.ts (hunk 2 above)
Comment (michaelbromley) [reply src/driver/mysql/MysqlQueryRunner.ts:1299]: Good catch. The value of `tableColumn.unsigned` is used inside the `isDefaultColumnWidth()` method, so it needs to be set prior to checking the width, which is why I moved it up. I missed the dependency on the `zerofill` value. I'll move the `zerofill` assignment above the `unsigned` assignment.

File: src/driver/mysql/MysqlQueryRunner.ts (hunk 3 above)
Comment (michaelbromley) [reply src/driver/mysql/MysqlQueryRunner.ts:1717]: Will do

**Comment by michaelbromley** [comment]
Those changes are done now 👍

**Comment by pleerock** [comment]
Thank you
